0000-ldap-sync-tool.md

```diff
@@ -0,0 +1,139 @@
+- Feature Name: `mgr-ldapsync`: LDAP sync tool
+- Start Date: 2019, Oct 28
+- RFC PR: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+The `mgr-ldapsync` is a tool that performs synchronisation of the
+users between Uyuni server and LDAP.
+
+# Motivation
+[motivation]: #motivation
+
+Uyuni server can authenticate users from LDAP via PAM. For this, users
+needs to be added to Uyuni with the flag "PAM authentication" and thus
+their credentials will be verified via LDAP directory. Synchronisation
+tool should transparently synchronise users from certain groups and/or
+roles in LDAP and reflect that in Uyuni server, overriding local
+changes.
+
+Often LDAP configuration at infrastructures is unpredictable. This
+tool should bring certain boundaries and over up most common LDAP
+setups for users and their roles management.
+
+# Detailed design
+[design]: #detailed-design
+
+The tool should work as an aggregator, collecting users from LDAP and
+Uyuni server synchronising them.
+
+## Functionality Aspects
+
+The following functions should be implemented:
+
+- Distinguish between local users in Uyuni and LDAP users.
+- Prevent UID clashes between the local users and LDAP users.
+- Prevent accidental modification or removal of the local user.
+- Flexible support various LDAP schemas, such as eDirectory, POSIX,
+  Active Directory etc.
+- Role management.
+- Dry run mode (preview only).
+
+## Configuration
+
+Configuration should be default in a separate file
+`/etc/rhn/ldapsync.conf`. Alternatively, user should be able to
+specify own path, in case the default one is not created or used.
+
+The main principle of the tool is to map specific roles and or groups
+in LDAP to the roles in the Uyuni server via the configuration. Adding
+or removing LDAP users from these groups or roles should reflect in
+Uyuni without additional manual changes in the Uyuni server.
+
+Configuration should be just a YAML file and the mapping should
+implement a relation between LDAP group to a set of roles from
+Uyuni. So each user in LDAP that is a member of such group would
+inherit these roles. For example:
+
+```yaml
+directory:
+  groups:
+    cn=sysop,ou=groups,dc=example,dc=com:
+      - activation_key_admin
+      - channel_admin
+      - config_admin
+```
+
+The configuration above would allow creating `groupOfNames` class in
+LDAP and add users via `member` attribute. As long as DN of that
+member is valid LDAP object and is not clashing with the local user in
+the Uyuni server, it should be synchronised. Similarly roles LDAP
+objects are configured as same as `groups` but via `roles`.
+
+## Frozen Users
```

Comment:
I think we do not need to define this at all. A local user has nothing in common with a ldap managed user.
We can document this as best practice but it should be obvious that you cannot login when your directory service is down and you have PAM auth configured.
On the other hand: when your directoy service is down you really have other problems as you may not login to your workstation.

Reply:
Local user still might have UID which is the same as in LDAP. Also you may want to freeze LDAP user as well for some reasons. Think of it as a guard to protect clashing UIDs or someone mistakenly adds a user in a wrong directory.

Reply:
> On the other hand: when your directoy service is down you really have other problems as you may not login to your workstation.

No credential caché on the workstation side? That's a given in Active Directory and Windows workstations.

Reply:
Well, while it is a valid point, I still believe this is beyond of the scope of this RFC and the tool itself. The tool is supposed to do only one job: preview the status, telling who is new and who is old and who is removed and then sync it, making sure no kittens were harmed. Done. How Uyuni is dealing _afterwards_ with the faulty/flaky LDAP through the PAM — is something else. Can we agree here?